go/tls/tls.go

```diff
@@ -0,0 +1,682 @@
+				v.Field(i).Set(reflect.New(structType.Field(i).Type.Elem()))
+				destination = v.Field(i).Elem()
+			}
+			offset, err = parseField(destination, data, offset, fieldInfo)
+			if err != nil {
+				return
+			}
+
+			// Remember any possible tls.Enum values encountered in case they are selectors
+			if structType.Field(i).Type.Kind() == enumType.Kind() {
+				enums[structType.Field(i).Name] = v.Field(i).Uint()
+			}
+
+		}
+		// Check that all selectors got a destination for their data
+		for selector, seen := range selectorSeen {
+			if !seen {
+				err = SyntaxError{selector + ": unhandled value for selector"}
+				return
+			}
+		}
+		return
+	case reflect.Array:
+		datalen := v.Len()
+
+		if datalen > len(rest) {
+			err = SyntaxError{prefix + "truncated array"}
+			return
+		}
+		inner := rest[:datalen]
+		offset += datalen
+		if fieldType.Elem().Kind() != reflect.Uint8 {
+			// Only byte/uint8 arrays are supported
+			err = StructuralError{prefix + "unsupported array type: " + v.Type().String()}
+			return
+		}
+		reflect.Copy(v, reflect.ValueOf(inner))
+		return
+
+	case reflect.Slice:
+		sliceType := fieldType
+		var varlen uint64
+		varlen, err = readVaruint(rest, info)
+		if err != nil {
+			return
+		}
+		datalen := int(varlen)
+		offset += int(info.count)
+		rest = rest[info.count:]
+
+		if datalen > len(rest) {
+			err = SyntaxError{prefix + "truncated slice"}
+			return
+		}
+		inner := rest[:datalen]
+		offset += datalen
+		if fieldType.Elem().Kind() == reflect.Uint8 {
+			// Fast version for []byte
+			v.Set(reflect.MakeSlice(sliceType, datalen, datalen))
+			reflect.Copy(v, reflect.ValueOf(inner))
+			return
+		}
+
+		v.Set(reflect.MakeSlice(sliceType, 0, datalen))
+		single := reflect.New(sliceType.Elem())
+		innerOffset := 0
+		for innerOffset < len(inner) {
+			innerOffset, err = parseField(single.Elem(), inner, innerOffset, nil)
+			if err != nil {
+				return
+			}
+			v.Set(reflect.Append(v, single.Elem()))
+		}
+		return
+
+	default:
+		err = StructuralError{fmt.Sprintf("%sunsupported type: %s of kind %s", prefix, fieldType, v.Kind())}
+		return
+	}
+}
+
+// Marshal returns the TLS encoding of val.
+func Marshal(val interface{}) ([]byte, error) {
+	return MarshalWithParams(val, "")
+}
+
+// MarshalWithParams returns the TLS encoding of val, and allows field
+// parameters to be specified for the top-level element.  The form
+// of the params is the same as the field tags.
+func MarshalWithParams(val interface{}, params string) ([]byte, error) {
+	info, err := fieldTagToFieldInfo(params, "")
+	if err != nil {
+		return nil, err
+	}
+	var out bytes.Buffer
+	v := reflect.ValueOf(val)
+	err = marshalField(&out, v, info)
+	if err != nil {
+		return nil, err
+	}
+	return out.Bytes(), err
+}
+
+func marshalField(out *bytes.Buffer, v reflect.Value, info *fieldInfo) (err error) {
+	var prefix string
+	if info != nil && len(info.name) > 0 {
+		prefix = info.name + ": "
+	}
```

Comment:
I've added a new `field` member to the structure to hold this information, so the prefix generation is now only done in the two `Error()` methods.